src/openrct2/network/Network.cpp

```diff
@@ -2529,43 +2578,71 @@ void Network::Server_Handle_TOKEN(NetworkConnection& connection, [[maybe_unused]
     Server_Send_TOKEN(connection);
 }
 
-void Network::Client_Handle_OBJECTS(NetworkConnection& connection, NetworkPacket& packet)
+void Network::Client_Handle_OBJECTS_LIST(NetworkConnection& connection, NetworkPacket& packet)
 {
     auto& repo = GetContext()->GetObjectRepository();
-    uint32_t size;
-    packet >> size;
-    log_verbose("client received object list, it has %u entries", size);
-    if (size > OBJECT_ENTRY_COUNT)
+
+    uint32_t index = 0;
+    uint32_t totalObjects = 0;
+    packet >> index >> totalObjects;
+
+    static constexpr uint32_t OBJECT_START_INDEX = 0;
+    if (index == OBJECT_START_INDEX)
+    {
+        _missingObjects.clear();
+    }
+
+    if (totalObjects > OBJECT_ENTRY_COUNT)
     {
         connection.SetLastDisconnectReason(STR_MULTIPLAYER_SERVER_INVALID_REQUEST);
         connection.Socket->Disconnect();
         log_warning("Server sent invalid amount of objects");
         return;
     }
-    std::vector<std::string> requested_objects;
-    for (uint32_t i = 0; i < size; i++)
+
+    if (totalObjects > 0)
     {
-        const char* name = reinterpret_cast<const char*>(packet.Read(8));
-        // Required, as packet has no null terminators.
-        std::string s(name, name + 8);
-        uint32_t checksum, flags;
+        char objectListMsg[256];
+        const uint32_t args[] = {
+            index + 1,
+            totalObjects,
+        };
+        format_string(objectListMsg, 256, STR_MULTIPLAYER_RECEIVING_OBJECTS_LIST, &args);
+
+        auto intent = Intent(WC_NETWORK_STATUS);
+        intent.putExtra(INTENT_EXTRA_MESSAGE, std::string{ objectListMsg });
+        intent.putExtra(INTENT_EXTRA_CALLBACK, []() -> void { gNetwork.Close(); });
+        context_open_intent(&intent);
+
+        char objectName[12]{};
+        std::memcpy(objectName, packet.Read(8), 8);
```

Comment:
Should we avoid `memcpy` here?

Also why does it have 3 extra positions?

Reply:
I could make it strictly 9 so the 9. would be the zero but it shouldn't matter much. Do you consider this worse than having a temporary std::string object? std::string::assign at the end is the same as a memcpy

Reply:
If it doesn't seem to be a critical piece of code, I'd just have a simple string assignment, I think it improves legibility and also makes it clear that the memcpy is not there for performance reasons

Reply:
In my eyes this clarifies this is a fixed length string, but I can revert this change if must.

Reply:
I don't want to make a fuss about it, your call

Reply:
Using std::string just didn't feel right for something rather trivial, but if anyone else has something to add please do.

Reply:
Why are you not doing `auto objectName = packet.Read(8);` ?

Reply:
Because its not null terminated